tests/integration/defs/test_unittests.py

```
@@ -63,6 +65,70 @@ def merge_attr(name, type_=int):
     base.write(output_file, encoding="UTF-8", xml_declaration=True)
 
 
+def _junit_culprits(xml_path: str) -> list[str]:
+    """Names of failed/errored tests in a JUnit report.
+
+    So a whole-case failure still names the specific culprit test(s) for
+    CI/triage attribution.
+    """
+    import xml.etree.ElementTree as ElementTree
+    try:
+        root = ElementTree.parse(xml_path).getroot()
```

Ruff (0.16.1)

[error] Using `xml` to parse untrusted data is known to be vulnerable to XML attacks; use `defusedxml` equivalents

(S314)